src/promptflow-devkit/promptflow/_sdk/operations/_flow_operations.py

```diff
@@ -979,41 +979,45 @@ def _resolve_requirements_txt(python_requirements, code):
         return None
 
     @staticmethod
-    def _resolve_signature(signature_overrides, entry, working_dir, language):
+    def _merge_signature(extracted, signature_overrides):
         if not signature_overrides:
             signature_overrides = {}
 
-        inspector_proxy = ProxyFactory().create_inspector_proxy(language=language)
-        if not inspector_proxy.is_flex_flow_entry(entry):
-            raise UserErrorException(f"Entry {entry} is not a valid entry for flow.")
-
-        # TODO: extract inits, and description?
-        entry_meta = inspector_proxy.get_entry_meta(entry=entry, working_dir=working_dir)
         signature = {}
         for key in ["inputs", "outputs", "init"]:
-            if key in entry_meta:
-                signature[key] = entry_meta[key]
+            if key in extracted:
+                signature[key] = extracted[key]
+            elif key in signature_overrides:
+                raise UserErrorException(f"Provided signature for {key}, which is not found in the entry.")
 
             if key not in signature_overrides:
                 continue
 
-            if set(signature[key].keys()) != set(signature_overrides[key].keys()):
+            if set(extracted[key].keys()) != set(signature_overrides[key].keys()):
                 raise UserErrorException(
-                    f"Provided signature of {key} for entry {entry} does not match the entry.\n"
-                    f"Ports with signature: {', '.join(signature_overrides[key].keys())}\n"
-                    f"Actual ports: {', '.join(signature[key].keys())}\n"
+                    f"Provided signature of {key} does not match the entry.\n"
+                    f"Ports from signature: {', '.join(signature_overrides[key].keys())}\n"
+                    f"Ports from entry: {', '.join(signature[key].keys())}\n"
                 )
 
             signature[key] = signature_overrides[key]
 
         return signature
 
+    @monitor_operation(activity_name="pf.flows._infer_signature", activity_type=ActivityType.INTERNALCALL)
+    def _infer_signature(self, entry: Callable, keep_entry: bool = False) -> Tuple[dict, Path]:
```

Comment:
we are free to make _infer_signature & save public now

Reply:
let me do this in a separate PR. @YingChen1996 and @lalala123123 should use this method as they should use internal call instead of public API?